massa-protocol-worker-2/src/handlers/operation_handler/propagation.rs

```diff
@@ -0,0 +1,148 @@
+    config: ProtocolConfig,
+    cache: SharedOperationCache,
+    operation_message_serializer: MessagesSerializer,
+}
+
+impl PropagationThread {
+    fn run(&mut self) {
+        let mut next_announce = std::time::Instant::now()
+            .checked_add(self.config.operation_announcement_interval.to_duration())
+            .expect("Can't init interval op propagation");
+        loop {
+            match self.internal_receiver.recv_deadline(next_announce) {
+                Ok(internal_message) => match internal_message {
+                    OperationHandlerCommand::AnnounceOperations(operations_ids) => {
+                        // Note operations as checked.
+                        {
+                            let mut cache_write = self.cache.write();
+                            for op_id in operations_ids.iter().copied() {
+                                cache_write.insert_checked_operation(op_id);
+                            }
+                        }
+                        self.operations_to_announce.extend(operations_ids);
+                        if self.operations_to_announce.len()
+                            > self.config.operation_announcement_buffer_capacity
+                        {
+                            self.announce_ops();
+                            next_announce = std::time::Instant::now()
+                                .checked_add(
+                                    self.config.operation_announcement_interval.to_duration(),
+                                )
+                                .expect("Can't init interval op propagation");
+                        }
+                    }
+                },
+                Err(RecvTimeoutError::Timeout) => {
+                    self.announce_ops();
+                    next_announce = std::time::Instant::now()
+                        .checked_add(self.config.operation_announcement_interval.to_duration())
+                        .expect("Can't init interval op propagation");
+                }
+                Err(RecvTimeoutError::Disconnected) => {
+                    return;
+                }
+            }
+        }
+    }
+
+    fn announce_ops(&mut self) {
+        // Quit if empty  to avoid iterating on nodes
+        if self.operations_to_announce.is_empty() {
+            return;
+        }
+        let operation_ids = mem::take(&mut self.operations_to_announce);
+        massa_trace!("protocol.protocol_worker.announce_ops.begin", {
+            "operation_ids": operation_ids
+        });
+        {
+            let mut cache_write = self.cache.write();
+            let peers: Vec<PeerId> = cache_write
+                .ops_known_by_peer
+                .iter()
+                .map(|(id, _)| id.clone())
+                .collect();
+            // Clean shared cache if peers do not exist anymore
+            for peer_id in peers {
+                if !self
+                    .active_connections
+                    .read()
```

Comment:
you can avoid call .read() in the for loop and call it before

Reply:
Done